File: android/src/main/kotlin/zoo/cityboy/appcenter_sdk/AppcenterSdkPlugin.kt
package zoo.cityboy.appcenter_sdk

import android.R
import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.microsoft.appcenter.AppCenter
import com.microsoft.appcenter.analytics.Analytics
import com.microsoft.appcenter.crashes.Crashes
import com.microsoft.appcenter.crashes.WrapperSdkExceptionManager
import com.microsoft.appcenter.distribute.Distribute
import com.microsoft.appcenter.distribute.DistributeListener
import com.microsoft.appcenter.distribute.ReleaseDetails
import com.microsoft.appcenter.distribute.UpdateAction
import com.microsoft.appcenter.distribute.UpdateTrack
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import zoo.cityboy.appcenter_sdk.analytics.AnalyticsApi
import zoo.cityboy.appcenter_sdk.appcenter.AppCenterApi
import zoo.cityboy.appcenter_sdk.appcenter.AppCenterConfig
import zoo.cityboy.appcenter_sdk.appcenter.LoggerLevel
import zoo.cityboy.appcenter_sdk.crashes.CrashesApi
import zoo.cityboy.appcenter_sdk.distribute.DistributeApi
import zoo.cityboy.appcenter_sdk.distribute.UpdateActionTap
import java.util.Date
import java.util.Locale


/** AppcenterSdkPlugin */
class AppcenterSdkPlugin: FlutterPlugin, ActivityAware, AppCenterApi, AnalyticsApi, CrashesApi,
    DistributeApi {

    private var events = mutableMapOf<String, Date>()
    private var mainActivity: Activity? = null
    private lateinit var preferences: SharedPreferences
    private var mEventChannel: EventChannel? = null
    private var mEvents: EventSink? = null
    private var uiThreadHandler: Handler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(p0: FlutterPluginBinding) {
        p0.applicationContext.getSharedPreferences("AppCenterSdkFlutter", Context.MODE_PRIVATE)
        events["onAttachedToEngine_start"] = Date()
        AppCenterApi.setUp(p0.binaryMessenger, this)
        AnalyticsApi.setUp(p0.binaryMessenger, this)
        CrashesApi.setUp(p0.binaryMessenger, this)
        DistributeApi.setUp(p0.binaryMessenger, this)
        events["onAttachedToEngine_done"] = Date()

        mEventChannel =
            EventChannel(p0.binaryMessenger, "zoo.cityboy.appcenter_sdk/distribute_events")
        mEventChannel?.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventSink) {
                Log.i("AppcenterSdk", "onListen $arguments -> $events")
                mEvents = events
            }

            override fun onCancel(arguments: Any?) {
                Log.i("AppcenterSdk", "onCancel -> $arguments")

                mEvents = null
            }
        })
    }

    override fun onDetachedFromEngine(p0: FlutterPluginBinding) {
        events["onDetachedFromEngine_start"] = Date()
        mEventChannel?.setStreamHandler(null)
        mEventChannel = null
        AppCenterApi.setUp(p0.binaryMessenger, null)
        AnalyticsApi.setUp(p0.binaryMessenger, null)
        CrashesApi.setUp(p0.binaryMessenger, null)
        DistributeApi.setUp(p0.binaryMessenger, null)
    }

    override fun onAttachedToActivity(p0: ActivityPluginBinding) {
        events["onAttachedToActivity_start"] = Date()
        mainActivity = p0.activity
        preferences =
            p0.activity.getSharedPreferences("AppCenterSdkFlutter", Context.MODE_PRIVATE);
        Distribute.checkForUpdate()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        events["onDetachedFromActivityForConfigChanges"] = Date()
        mainActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(p0: ActivityPluginBinding) {

        events["onReattachedToActivityForConfigChanges"] = Date()
        mainActivity = p0.activity
        preferences =
            p0.activity.getSharedPreferences("AppCenterSdkFlutter", Context.MODE_PRIVATE);
        
    }

    override fun onDetachedFromActivity() {
        mainActivity = null
        events["onDetachedFromActivity"] = Date()
    }

    // App Center


    override fun start(config: AppCenterConfig) {
        events["AppCenterStart"] = Date()
        Distribute.setEnabled(config.distributeEnabled)
        Distribute.setListener( AppCenterDistributeListener())


        mainActivity?.let { it ->
            if (!AppCenter.isConfigured()) {


//                    object : DistributeListener {
//                        override fun onNoReleaseAvailable(activity: Activity?) {
//                            Log.i("AppcenterSdk", "Distribute: onNoReleaseAvailable")
//                            uiThreadHandler.postDelayed({ mEvents?.success(null) }, 1000)
//                        }
//                        override fun onReleaseAvailable(
//                            activity: Activity?,
//                            releaseDetails: ReleaseDetails?
//                        ): Boolean {
//                            releaseDetails?.let {detail ->
//                                try{
//                                    val converted = mapOf<String, Any?>(
//                                        "id" to detail.id.toInt(),
//                                        "version" to detail.version.toInt(),
//                                        "shortVersion" to detail.shortVersion.toString(),
//                                        "size" to detail.size.toInt(),
//                                        "releaseNotes" to detail.releaseNotes?.toString(),
//                                        "releaseNotesUrl" to detail.releaseNotesUrl?.toString(),
//                                        "minApiLevel" to Build.VERSION.SDK_INT,
//                                        "downloadUrl" to detail.downloadUrl.toString(),
//                                        "isMandatoryUpdate" to detail.isMandatoryUpdate,
//                                        "releaseHash" to detail.releaseHash.toString(),
//                                        "distributionGroupId" to
//                                                detail.distributionGroupId.toString(),
//                                    )
//                                    Log.i("AppcenterSdk", "[kt] Distribute: onReleaseAvailable $converted")
//
//                                    uiThreadHandler.postDelayed( { mEvents?.success(converted) },1000)
//
//                                } catch (e: Exception){
//                                    Log.i("AppcenterSdk", "error: ${e.message}")
//                                    uiThreadHandler.postDelayed({ mEvents?.success(null) }, 1000)
//
//                                    return false
//                                }
//
//                            } ?: {
//                                Log.i("AppcenterSdk", "Distribute: onReleaseAvailable $releaseDetails")
//                                uiThreadHandler.postDelayed( { mEvents?.success(null) },1000)
//                            }
//
//                            return true
//                        }
//                    }
//                )

                Analytics.setTransmissionInterval(config.transmissionInterval.toInt())
                AppCenter.start(it.application, config.secret, Analytics::class.java, Crashes::class.java, Distribute::class.java)
                events["AppCenterStart_done"] = Date()


            } else {
                Log.i("AppcenterSdk", "AppCenter is configured")
            }
            setDistributeEnabled(config.crashEnabled,{_->})
            crashesSetEnabled(config.crashEnabled,{_->})
            analyticsSetEnabled(config.analyticsEnabled,{_->})
            setLogLevel(config.logLevel)

            Log.i("AppcenterSdk", "start: ${AppCenter.isConfigured()}")
        }
    }


    override fun setEnabled(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
        Log.i("AppcenterSdk", "setEnabled: $enabled")
        AppCenter.setEnabled(enabled).thenAccept { callback(Result.success(Unit)) }
    }

    override fun isEnabled(callback: (Result<Boolean>) -> Unit) {
        try {
            AppCenter.isEnabled().thenAccept { value ->
                Log.i("AppcenterSdk", "isEnabled: $value")
                callback(Result.success(value))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }

    override fun isConfigured(): Boolean = AppCenter.isConfigured()

    override fun getInstallId(callback: (Result<String>) -> Unit) {
        try {
            AppCenter.getInstallId().thenAccept { uuid ->
                Log.i("AppcenterSdk", "getInstallId: $uuid")

                callback(Result.success(uuid.toString()))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }

    override fun isRunningInAppCenterTestCloud(): Boolean = AppCenter.isRunningInAppCenterTestCloud()

    override fun setUserId(userId: String, callback: (Result<Unit>) -> Unit) {
        try{
            AppCenter.setUserId(userId);
            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }

    override fun setCountryCode(countryCode: String, callback: (Result<Unit>) -> Unit) {
        try{
            AppCenter.setCountryCode(countryCode);
            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }

    override fun getSdkVersion(): String {
        return AppCenter.getSdkVersion()
    }

    override fun setNetworkRequestsAllowed(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
        try{
            AppCenter.setNetworkRequestsAllowed(enabled);
            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }

    override fun isNetworkRequestsAllowed(): Boolean {
        return AppCenter.isNetworkRequestsAllowed()
    }

    override fun setLogLevel(level: LoggerLevel) {
        Log.i("AppcenterSdk", "setLogLevel: $level")
        AppCenter.setLogLevel(level.ordinal)
    }

    // App Center Analytics
    override fun trackEvent(name: String, properties: Map<String, String>?, flags: Long?) {
        Log.i("AppcenterSdk", "trackEvent: name: $name properties: $properties flags: $flags")

        if (flags != null) {
            Analytics.trackEvent(name, properties, flags.toInt())
        } else {
            Analytics.trackEvent(name, properties)
        }
    }

    override fun trackPage(name: String, properties: Map<String, String>?) {
        Analytics.trackEvent(name, properties)
    }

    override fun pause() {
        Analytics.pause()
    }

    override fun resume() {
        Analytics.resume()
    }

    override fun analyticsSetEnabled(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
        try {
            Analytics.setEnabled(enabled).thenAccept {
                Log.i("AppcenterSdk", "analyticsSetEnabled: $enabled")

                callback(Result.success(Unit))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }

    override fun analyticsIsEnabled(callback: (Result<Boolean>) -> Unit) {
        try {
            Analytics.isEnabled().thenAccept { value ->
                Log.i("AppcenterSdk", "analyticsIsEnabled: $value")
                callback(Result.success(value))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }

    override fun enableManualSessionTracker() {
        Analytics.enableManualSessionTracker()
    }

    override fun startSession() {
        Analytics.startSession()
    }

    override fun setTransmissionInterval(seconds: Long): Boolean = Analytics.setTransmissionInterval(seconds.toInt())

    // App Center Crashes
    override fun generateTestCrash() {
        Crashes.generateTestCrash()
    }

    override fun hasReceivedMemoryWarningInLastSession(callback: (Result<Boolean>) -> Unit) {
        try {
            Crashes.hasReceivedMemoryWarningInLastSession()
                .thenAccept { value -> callback(Result.success(value)) }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }

    override fun hasCrashedInLastSession(callback: (Result<Boolean>) -> Unit) {
        try {
            Crashes.hasCrashedInLastSession()
                .thenAccept { value ->
                    Log.i("AppcenterSdk", "hasCrashedInLastSession: $value")
                    callback(Result.success(value))
                }
        } catch (e: Exception)  {
            callback(Result.failure(e))
        }

    }

    override fun crashesSetEnabled(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
        try {
            Crashes.setEnabled(enabled).thenAccept {
                Log.i("AppcenterSdk", "crashesSetEnabled: $enabled")
                callback(Result.success(Unit))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }

    override fun crashesIsEnabled(callback: (Result<Boolean>) -> Unit) {
        try {
            Crashes.isEnabled().thenAccept { value ->
                Log.i("AppcenterSdk", "crashesIsEnabled: $value")
                callback(Result.success(value))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }

    override fun trackException(
        message: String,
        type: String?,
        stackTrace: String?,
        properties: Map<String, String>?
    ) {
        val exceptionModel = com.microsoft.appcenter.crashes.ingestion.models.Exception()
        exceptionModel.message = message
        exceptionModel.type = type
        exceptionModel.stackTrace = stackTrace
        exceptionModel.wrapperSdkName = "appcenter.flutter"
        WrapperSdkExceptionManager.trackException(exceptionModel, properties, null)
    }
    // App Center Crashes
    

    // App Center Distribution

    override fun setDistributeEnabled(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
        try {
            Distribute.setEnabled(enabled).thenAccept {
                Log.i("AppcenterSdk", "setDistributeEnabled: $enabled")
                callback(Result.success(Unit))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }


    override fun setDistributeDebugEnabled(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
       try {
           Log.i("AppcenterSdk", "setEnabledForDebuggableBuild: $enabled")
           Distribute.setEnabledForDebuggableBuild(enabled)
           callback(Result.success(Unit))
       } catch (e: Exception){
           callback(Result.failure(e))
       }

    }

    override fun isDistributeEnabled(callback: (Result<Boolean>) -> Unit) {
        try {
            Distribute.isEnabled().thenAccept { value ->
                Log.i("AppcenterSdk", "isDistributeEnabled: $value")
                callback(Result.success(value))
            }
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }


    override fun disableAutomaticCheckForUpdate(callback: (Result<Unit>) -> Unit) {
        try {
            Distribute.disableAutomaticCheckForUpdate()
            Log.i("AppcenterSdk", "disableAutomaticCheckForUpdate")

            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }

    }
    

    override fun checkForUpdates(callback: (Result<Unit>) -> Unit) {
        try {
            Distribute.checkForUpdate()
            Log.i("AppcenterSdk", "checkForUpdate")
            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }

    override fun ifExistsReleaseDetails(): zoo.cityboy.appcenter_sdk.distribute.ReleaseDetails? {
        return null
    }

    override fun notifyDistributeUpdateAction(updateAction:UpdateActionTap, callback: (Result<Unit>) -> Unit) {
        try{
            val action = when (updateAction){
                UpdateActionTap.UPDATE -> com.microsoft.appcenter.distribute.UpdateAction.UPDATE
                UpdateActionTap.POSTPONE -> com.microsoft.appcenter.distribute.UpdateAction.POSTPONE
            }
            Distribute.notifyUpdateAction(action)
            Log.i("AppcenterSdk", "notifyUpdateAction: $updateAction")
            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }

    override fun handleDistributeUpdateAction(updateAction: UpdateActionTap, callback: (Result<Unit>) -> Unit) {
        try{
            val action = when (updateAction){
                UpdateActionTap.UPDATE -> com.microsoft.appcenter.distribute.UpdateAction.UPDATE
                UpdateActionTap.POSTPONE -> com.microsoft.appcenter.distribute.UpdateAction.POSTPONE
            }
            Distribute.notifyUpdateAction(action)
            Log.i("AppcenterSdk", "notifyUpdateAction: $updateAction")
            callback(Result.success(Unit))
        } catch (e: Exception){
            callback(Result.failure(e))
        }
    }
}

class AppCenterDistributeListener : DistributeListener {
    override fun onReleaseAvailable(activity: Activity, releaseDetails: ReleaseDetails): Boolean {
        val releaseNotes = releaseDetails.releaseNotes
        val custom =  releaseNotes != null

            val dialogBuilder = MaterialAlertDialogBuilder(activity)
                .setIcon(android.R.drawable.stat_sys_download_done)
            .setTitle(
                String.format(
                    activity.getString(com.microsoft.appcenter.distribute.R.string.appcenter_distribute_update_dialog_title),
                    releaseDetails.shortVersion
                )
            )
            .setMessage(releaseNotes)
            .setPositiveButton(com.microsoft.appcenter.distribute.R.string.appcenter_distribute_update_dialog_download,
                DialogInterface.OnClickListener { dialog, which ->
                    Distribute.notifyUpdateAction(
                        UpdateAction.UPDATE
                    )
                })
            .setCancelable(false)
            if (!releaseDetails.isMandatoryUpdate) {
                dialogBuilder.setNegativeButton(com.microsoft.appcenter.distribute.R.string.appcenter_distribute_update_dialog_postpone,
                    DialogInterface.OnClickListener { dialog, which ->
                        Distribute.notifyUpdateAction(
                            UpdateAction.POSTPONE
                        )
                    })
            }
            dialogBuilder.setNeutralButton(com.microsoft.appcenter.distribute.R.string.appcenter_distribute_update_dialog_view_release_notes,
                DialogInterface.OnClickListener { dialogInterface, i ->
                    activity.startActivity(
                        Intent(Intent.ACTION_VIEW, releaseDetails.releaseNotesUrl)
                    )
                })
            dialogBuilder.show()

        return custom
    }

    override fun onNoReleaseAvailable(activity: Activity) {
        Toast.makeText(
            activity,
            "No updates available",
            Toast.LENGTH_LONG
        ).show()
    }
}

File: ios/Classes/AppcenterSdkPlugin.swift
import Flutter
import UIKit
import AppCenter
import AppCenterAnalytics
import AppCenterCrashes
#if canImport(AppCenterDistribute)
    import AppCenterDistribute
#endif
public class AppcenterSdkPlugin: NSObject, FlutterPlugin, AppCenterApi {
    
    
       
    var mEvents: FlutterEventSink?
    var mEventChannel: FlutterEventChannel?
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let binaryMessenger = registrar.messenger()
        let appCenterPlugin = AppcenterSdkPlugin()
        
        AppCenterApiSetup.setUp(binaryMessenger: binaryMessenger, api: appCenterPlugin)
        AnalyticsApiSetup.setUp(binaryMessenger: binaryMessenger, api: appCenterPlugin)
        CrashesApiSetup.setUp(binaryMessenger: binaryMessenger, api: appCenterPlugin)
        #if canImport(AppCenterDistribute)
            DistributeApiSetup.setUp(binaryMessenger: binaryMessenger, api: appCenterPlugin)
        #endif
        
        
        appCenterPlugin.mEventChannel = FlutterEventChannel(name: "zoo.cityboy.appcenter_sdk/distribute_events", binaryMessenger: binaryMessenger)
        appCenterPlugin.mEventChannel?.setStreamHandler(appCenterPlugin)
        
    }
    
    // AppCenter
    
    
    func setUserId(userId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        AppCenter.userId = userId
        completion(.success(Void()))
    }
    
    func setCountryCode(countryCode: String, completion: @escaping (Result<Void, Error>) -> Void) {
        AppCenter.countryCode = countryCode
    }
    
    func getSdkVersion() throws -> String {
        return AppCenter.sdkVersion
    }
    
    func setNetworkRequestsAllowed(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        AppCenter.networkRequestsAllowed = enabled
    }
    
    func isNetworkRequestsAllowed() throws -> Bool {
        return AppCenter.networkRequestsAllowed
    }
//    func start(secret: String, usePrivateTrack: Bool) {
    func start(config: AppCenterConfig) throws {
        NSLog("Appcenter: start config \(config)")
    
        if (!AppCenter.isConfigured){
            Analytics.transmissionInterval = UInt(config.transmissionInterval)
            
            var services = [Analytics.self, Crashes.self]
            #if canImport(AppCenterDistribute)
                if config.usePrivateTrack {
                    Distribute.updateTrack = .private
                } else {
                    Distribute.updateTrack = .public
                }
                services.append(Distribute.self)
            #endif
            AppCenter.start(withAppSecret: config.secret, services: services)
            
            AppCenter.logLevel = config.logLevel.toLogLevel()
            Analytics.enabled = config.analyticsEnabled
            Crashes.enabled = config.crashEnabled
            #if canImport(AppCenterDistribute)
                Distribute.enabled = config.distributeEnabled
            #endif
            NSLog("Appcenter: start new configuration")
            
        } else {
            NSLog("Appcenter: start configured")
        }
    }
    
    
    func setEnabled(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        AppCenter.enabled = enabled
        NSLog("Appcenter: setEnabled \(enabled)")
        completion(.success(Void()))
    }
    
    func isEnabled(completion: @escaping (Result<Bool, Error>) -> Void) {
        NSLog("Appcenter: isEnabled \(AppCenter.enabled)")
        completion(.success(AppCenter.enabled))
    }
    
    func isConfigured() -> Bool {
        NSLog("Appcenter: isConfigured \(AppCenter.isConfigured)")
        return AppCenter.isConfigured
    }
    
    func getInstallId(completion: @escaping (Result<String, Error>) -> Void) {
        NSLog("Appcenter: getInstallId \(AppCenter.installId.uuidString)")
        completion(.success(AppCenter.installId.uuidString))
    }
    
    func isRunningInAppCenterTestCloud() -> Bool {
        NSLog("Appcenter: isRunningInAppCenterTestCloud \(AppCenter.isRunningInAppCenterTestCloud)")
        return AppCenter.isRunningInAppCenterTestCloud
    }
    
    func setLogLevel(level: LoggerLevel) throws {
        NSLog("Appcenter: setLogLevel \(level)")
        AppCenter.logLevel = level.toLogLevel()
    }

// AppCenter
}
extension LoggerLevel{
    func toLogLevel() -> LogLevel {
            switch self {
            case .verbose: return LogLevel.verbose
            case .debug: return LogLevel.debug
            case .info: return LogLevel.info
            case .warning: return LogLevel.warning
            case .error: return LogLevel.error
            case .fatal: return LogLevel.assert
            case .none: return LogLevel.none
            }
        }
}
extension AppcenterSdkPlugin: CrashesApi {
    func generateTestCrash() {
        NSLog("Appcenter: generateTestCrash")
        Crashes.generateTestCrash()
    }
    
    func hasReceivedMemoryWarningInLastSession(completion: @escaping (Result<Bool, Error>) -> Void) {
        NSLog("Appcenter: hasReceivedMemoryWarningInLastSession \(Crashes.hasReceivedMemoryWarningInLastSession)")
        completion(.success(Crashes.hasReceivedMemoryWarningInLastSession))
    }
    
    func hasCrashedInLastSession(completion: @escaping (Result<Bool, Error>) -> Void) {
        NSLog("Appcenter: hasCrashedInLastSession \(Crashes.hasCrashedInLastSession)")
        completion(.success(Crashes.hasCrashedInLastSession))
    }
    
    func crashesSetEnabled(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        NSLog("Appcenter: crashesSetEnabled \(enabled)")
        Crashes.enabled = enabled
        completion(.success(Void()))
    }
    
    func crashesIsEnabled(completion: @escaping (Result<Bool, Error>) -> Void) {
        NSLog("Appcenter: crashesIsEnabled \(Crashes.enabled)")
        completion(.success(Crashes.enabled))
    }
    
    func trackException(message: String, type: String?, stackTrace: String?, properties: [String : String]?) {
        let exceptionModel = MSACWrapperExceptionModel()
        exceptionModel.message = message
        exceptionModel.type = type
        exceptionModel.stackTrace = stackTrace
        exceptionModel.wrapperSdkName = "appcenter.flutter"
        
        NSLog("Appcenter: trackException \(exceptionModel)")
        Crashes.trackException(exceptionModel, properties: properties, attachments: nil)
    }
}
extension AppcenterSdkPlugin: AnalyticsApi {
    
    func trackPage(name: String, properties: [String : String]?) throws {
        NSLog("Appcenter: trackPage \(name) -> \(String(describing: properties))")
        Analytics.trackEvent(name, withProperties: properties)
    }
    
    func trackEvent(name: String, properties: [String : String]?, flags: Int64?)throws {
        NSLog("Appcenter: trackEvent \(name) -> \(String(describing: properties)) flags: \(String(describing: flags))")
        if(flags == nil){
            Analytics.trackEvent(name, withProperties: properties)
        }else{
            Analytics.trackEvent(name, withProperties: properties, flags: Flags.init(rawValue: UInt.init(flags!)))
        }
    }
    
    func pause() {
        NSLog("Appcenter: pause ")
        Analytics.pause()
    }
    
    func resume() {
        NSLog("Appcenter: resume")
        Analytics.resume()
    }
    
    func analyticsSetEnabled(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        Analytics.enabled = enabled
        NSLog("Appcenter: analyticsSetEnabled \(enabled) ")
        completion(.success(Void()))
    }
    
    func analyticsIsEnabled(completion: @escaping (Result<Bool, Error>) -> Void) {
        NSLog("Appcenter: analyticsIsEnabled \(Analytics.enabled) ")
        completion(.success(Analytics.enabled))
    }
    
    func enableManualSessionTracker() {
        NSLog("Appcenter: enableManualSessionTracker")
        Analytics.enableManualSessionTracker()
    }
    
    func startSession() {
        NSLog("Appcenter: startSession")
        Analytics.startSession()
    }
    
    func setTransmissionInterval(seconds: Int64) throws -> Bool {
        NSLog("Appcenter: transmissionInterval \(seconds)")
        Analytics.transmissionInterval = UInt.init(seconds)
        return true
    }
}

extension AppcenterSdkPlugin: UIApplicationDelegate{
//    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
//        // Forward the URL.
//        return Distribute.open(url);
//      }
    public func applicationDidBecomeActive(_ application: UIApplication) {
        NSLog("Appcenter: applicationDidBecomeActive")
    }

    public func applicationWillTerminate(_ application: UIApplication) {
        NSLog("Appcenter: applicationWillTerminate")
    }

    public func applicationWillResignActive(_ application: UIApplication) {
        NSLog("Appcenter: applicationWillResignActive")
    }

    public func applicationDidEnterBackground(_ application: UIApplication) {
        NSLog("Appcenter: applicationDidEnterBackground")
    }

    public func applicationWillEnterForeground(_ application: UIApplication) {
        NSLog("Appcenter: applicationWillEnterForeground")
    }
}
extension AppcenterSdkPlugin: FlutterStreamHandler {
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
      NSLog("Appcenter: distribute onListen: arguments \(String(describing: arguments)) events: \(String(describing: events))")
    self.mEvents = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.mEvents = nil
    return nil
  }
}
#if canImport(AppCenterDistribute)
extension AppcenterSdkPlugin: DistributeApi{
    func setDistributeEnabled(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        #if canImport(AppCenterDistribute)
            NSLog("Appcenter: setDistributeEnabled \(enabled)")
            Distribute.enabled = enabled
            completion(.success(Void()))
        #else
            NSLog("Appcenter: [distribute not available] setDistributeEnabled \(enabled)")
            completion(.success(Void()))
        #endif
       
    }
    
    func notifyDistributeUpdateAction(updateAction: UpdateActionTap, completion: @escaping (Result<Void, Error>) -> Void) {
        #if canImport(AppCenterDistribute)
            NSLog("Appcenter: notifyDistributeUpdateAction \(updateAction)")
            if (updateAction == UpdateActionTap.update){
               Distribute.notify(.update)
            } else if (updateAction == UpdateActionTap.postpone){
               Distribute.notify(.postpone)
            }
        #else
        NSLog("Appcenter: [distribute not available] notifyDistributeUpdateAction \(updateAction)")
        #endif
    }
    
    func handleDistributeUpdateAction(updateAction: UpdateActionTap, completion: @escaping (Result<Void, Error>) -> Void) {
        #if canImport(AppCenterDistribute)
            NSLog("Appcenter: handleDistributeUpdateAction \(updateAction)")
            completion(.success(Void()))
        #else
        NSLog("Appcenter: [distribute not available] handleDistributeUpdateAction \(updateAction)")
        #endif
    }
    
    func setDistributeDebugEnabled(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        NSLog("Appcenter: [distribute not available] setDistributeDebugEnabled \(enabled)")
        completion(.success(Void()))
    }
    
    func isDistributeEnabled(completion: @escaping (Result<Bool, Error>) -> Void) {
        #if canImport(AppCenterDistribute)
            NSLog("Appcenter: isDistributeEnabled \(Distribute.enabled)")
            completion(.success(Distribute.enabled))
        #else
            NSLog("Appcenter: [distribute not available] isDistributeEnabled \(Distribute.enabled)")
            completion(.success(false))
        #endif
    }
    
    func disableAutomaticCheckForUpdate(completion: @escaping (Result<Void, Error>) -> Void) {
        #if canImport(AppCenterDistribute)
            NSLog("Appcenter: [distribute not available] disableAutomaticCheckForUpdate")
            Distribute.disableAutomaticCheckForUpdate()
        #else
            NSLog("Appcenter: [distribute not available] disableAutomaticCheckForUpdate")
        #endif
        completion(.success(Void()))
    }
    
    func checkForUpdates(completion: @escaping (Result<Void, Error>) -> Void) {
        #if canImport(AppCenterDistribute)
            NSLog("Appcenter: checkForUpdates")
            Distribute.checkForUpdate()
        #else
            NSLog("Appcenter: [distribute not available] checkForUpdates")
        #endif
        completion(.success(Void()))
    }
    
    func ifExistsReleaseDetails() throws -> ReleaseDetails? {
        return nil
    }
}

extension AppcenterSdkPlugin: DistributeDelegate{
    
    private func distribute(_ distribute: Distribute, releaseAvailableWith details: ReleaseDetails) -> Bool {
        NSLog("Appcenter: distribute details: \(details)")
        // mEvents = details
//        if UserDefaults.standard.bool(forKey: kSASCustomizedUpdateAlertKey) {
//
//              // Show a dialog to the user where they can choose if they want to update.
//              let alertController = UIAlertController(title: NSLocalizedString("distribute_alert_title", tableName: "Sasquatch", comment: ""),
//                      message: NSLocalizedString("distribute_alert_message", tableName: "Sasquatch", comment: ""),
//                      preferredStyle: .alert)
//
//              // Add a "Yes"-Button and call the notifyUpdateAction-callback with .update
//              alertController.addAction(UIAlertAction(title: NSLocalizedString("distribute_alert_yes", tableName: "Sasquatch", comment: ""), style: .cancel) { _ in
//                Distribute.notify(.update)
//              })
//
//              // Add a "No"-Button and call the notifyUpdateAction-callback with .postpone
//              alertController.addAction(UIAlertAction(title: NSLocalizedString("distribute_alert_no", tableName: "Sasquatch", comment: ""), style: .default) { _ in
//                Distribute.notify(.postpone)
//              })
//
//              // Show the alert controller.
//              self.topMostViewController()?.present(alertController, animated: true)
//              return true
//            }
//            return false
        return true;
    }
    public func distributeNoReleaseAvailable(_ distribute: Distribute) {
        NSLog("Appcenter: distributeNoReleaseAvailable")
    }
    public func distributeWillExitApp(_ distribute: Distribute) {
        NSLog("Appcenter: distributeWillExitApp")
    }
    
}
#endif
